module-gui/gui/widgets/text/core/lines/SingleTextLine.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2017-2022, Mudita Sp. z.o.o. All rights reserved.
+// For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
+
+#include "SingleTextLine.hpp"
+
+namespace gui
+{
+    SingleTextLine::SingleTextLine(BlockCursor &localCursor, unsigned int maxWidth, TextEllipsis ellipsis)
+        : TextLine(maxWidth), ellipsis(ellipsis)
+    {
+        do {
+            if (!localCursor) { // cursor is faulty
+                lineEnd = true;
+                return;
+            }
+
+            if (localCursor.atEnd()) {
+                lineEnd = true;
+                return;
+            }
+
+            // take text we want to show
+            auto text = localCursor.getUTF8Text();
+
+            if (text.length() == 0 && !localCursor.atEnd()) {
+                setLineStartConditions(localCursor.getBlockNumber(), localCursor.getPosition());
+                localCursor.goToNextBlock();
+                continue;
+            }
+
+            auto textFormat = localCursor->getFormat();
+            if (textFormat->getFont() == nullptr) {
+                lineEnd = true;
+                return;
+            }
+
+            // check if max provided width is enough to enter one char at least
+            if (maxWidth < textFormat->getFont()->getCharPixelWidth(text[0])) {
+                lineEnd = true;
+                return;
+            }
+
+            auto signsCountToShow = calculateSignsToShow(localCursor, text, maxWidth - widthUsed);
+
+            // create item for show and update Line data
+            auto item = buildUITextPart(textToPrint(signsCountToShow, text), textFormat);
+            shownLetterCount += signsCountToShow;
+            widthUsed += item->widgetArea.w;
+            heightUsed = std::max(heightUsed, item->widgetArea.h);
+            lineContent.emplace_back(item);
+
+            setLineStartConditions(localCursor.getBlockNumber(), localCursor.getPosition());
+            localCursor += signsCountToShow;
+
+            if (signsCountToShow < text.length()) {
+                lineEnd = true;
+                return;
+            }
+
+        } while (true);
+    }
+
+    bool SingleTextLine::checkLeftEndCondition(BlockCursor &localCursor)
+    {
+        return localCursor.getBlockNumber() != 0 || localCursor.getPosition() != 0;
+    }
+
+    bool SingleTextLine::checkRightEndCondition(BlockCursor &localCursor, unsigned int signsCountToShow)
+    {
+        BlockCursor moveCursor(localCursor.getDocument(),
+                               localCursor.getPosition(),
+                               localCursor.getBlockNumber(),
+                               localCursor->getFormat()->getFont());
+        moveCursor += (signsCountToShow);
+
+        return !moveCursor.atEnd();
+    }
+
+    // This method need correction - it does not handle block properly but there is no use case in system to
+    // handle it for now.
```

Comment:
In what way the block is not handled properly? 
Could you please elaborate?

Reply:
There is an issue that you need to know in what style you should print ellipsis -> it probably would have to use style from corresponding block and all is written in a way that it should but there is a lot of corner cases that does not need to be resolved right now (or even ever -> I do not see clear use case for it).